File: ptcgp/engine/play_card.py
"""Functions for playing cards from hand."""
from __future__ import annotations

from typing import Optional

from ptcgp.cards.card import Card
from ptcgp.cards.database import get_card
from ptcgp.cards.types import CardKind, Stage
from ptcgp.engine.actions import SlotRef
from ptcgp.engine.slot_utils import get_slot
from ptcgp.engine.state import GameState, PlayerState, PokemonSlot


def _take_from_hand(
    player: PlayerState,
    hand_index: int,
    expected_kind: Optional[CardKind] = None,
) -> Card:
    """Validate ``hand_index`` and return the card at that slot.

    Does NOT mutate the hand. Use after validation succeeds via ``player.hand.pop``.
    """
    if hand_index < 0 or hand_index >= len(player.hand):
        raise ValueError(
            f"Invalid hand_index {hand_index}: hand has {len(player.hand)} cards"
        )
    card = get_card(player.hand[hand_index])
    if expected_kind is not None and card.kind != expected_kind:
        raise ValueError(
            f"Card {card.name!r} is not a {expected_kind.name} card (kind: {card.kind})"
        )
    return card


def play_basic(state: GameState, hand_index: int, bench_slot: int) -> GameState:
    """Play a Basic Pokemon from hand to a specific bench slot."""
    state = state.copy()
    player = state.players[state.current_player]

    card = _take_from_hand(player, hand_index, CardKind.POKEMON)
    if card.stage != Stage.BASIC:
        raise ValueError(f"Card {card.name!r} is not a Basic Pokemon (stage: {card.stage})")

    if bench_slot < 0 or bench_slot >= len(player.bench):
        raise ValueError(f"Invalid bench_slot {bench_slot}")
    if player.bench[bench_slot] is not None:
        raise ValueError(f"Bench slot {bench_slot} is already occupied")

    player.hand.pop(hand_index)
    player.bench[bench_slot] = PokemonSlot(
        card_id=card.id,
        current_hp=card.hp,
        max_hp=card.hp,
        turns_in_play=0,
    )
    return state


def play_item(
    state: GameState,
    hand_index: int,
    target: Optional[SlotRef] = None,
    extra_hand_index: Optional[int] = None,
) -> GameState:
    """Play an Item card: discard it and resolve its trainer effect text.

    ``target`` (if provided) is passed through as ``target_ref`` to the effect
    handler — used by Potion and other ``heal 1 of your Pokémon`` items.
    ``extra_hand_index`` lets items like Rare Candy reference a second card
    in hand (the Stage 2 card to be consumed).
    """
    state = state.copy()
    player = state.players[state.current_player]

    card = _take_from_hand(player, hand_index, CardKind.ITEM)
    # Capture the extra card's ID (if any) BEFORE we pop Rare Candy from hand,
    # so the indices in the player's hand stay stable while we look it up.
    extra_card_id: Optional[str] = None
    if extra_hand_index is not None and 0 <= extra_hand_index < len(player.hand):
        extra_card_id = player.hand[extra_hand_index]

    player.hand.pop(hand_index)
    player.discard.append(card.id)

    # After popping Rare Candy, the stage 2 card's hand index may shift down
    # by one if it came after Rare Candy in the original hand.
    adjusted_extra_idx: Optional[int] = None
    if extra_hand_index is not None:
        if extra_hand_index > hand_index:
            adjusted_extra_idx = extra_hand_index - 1
        else:
            adjusted_extra_idx = extra_hand_index

    if card.trainer_effect_text or card.trainer_handler:
        from ptcgp.effects.apply import apply_effects
        extra: dict = {}
        if extra_card_id is not None:
            extra["evo_card_id"] = extra_card_id
            extra["evo_hand_index"] = adjusted_extra_idx
        state = apply_effects(
            state,
            card.trainer_effect_text,
            acting_player=state.current_player,
            target_ref=target,
            extra=extra,
            handler_str=card.trainer_handler,
            cached_effects=card.cached_trainer_effects,
        )
    return state


def play_supporter(
    state: GameState,
    hand_index: int,
    target: Optional[SlotRef] = None,
) -> GameState:
    """Play a Supporter card: discard, mark flag, resolve its effect text."""
    state = state.copy()
    player = state.players[state.current_player]

    card = _take_from_hand(player, hand_index, CardKind.SUPPORTER)
    player.hand.pop(hand_index)
    player.discard.append(card.id)
    player.has_played_supporter = True

    if card.trainer_effect_text or card.trainer_handler:
        from ptcgp.effects.apply import apply_effects
        state = apply_effects(
            state,
            card.trainer_effect_text,
            acting_player=state.current_player,
            target_ref=target,
            handler_str=card.trainer_handler,
            cached_effects=card.cached_trainer_effects,
        )
    return state


def attach_tool(state: GameState, hand_index: int, target: SlotRef) -> GameState:
    """Attach a Tool card to a Pokemon and apply its passive effect."""
    state = state.copy()
    player = state.players[state.current_player]

    card = _take_from_hand(player, hand_index, CardKind.TOOL)

    slot = get_slot(state, target)
    if slot is None:
        raise ValueError(f"No Pokemon in target slot {target}")
    if slot.tool_card_id is not None:
        raise ValueError(f"Target Pokemon already has a tool attached ({slot.tool_card_id!r})")

    player.hand.pop(hand_index)
    slot.tool_card_id = card.id

    if card.trainer_effect_text or card.trainer_handler:
        from ptcgp.effects.apply import apply_effects
        state = apply_effects(
            state,
            card.trainer_effect_text,
            acting_player=state.current_player,
            target_ref=target,
            handler_str=card.trainer_handler,
            cached_effects=card.cached_trainer_effects,
        )
    return state

File: ptcgp/src/engine/play_card.rs
//! Functions for playing cards from hand.
//!
//! Ported from `ptcgp/engine/play_card.py`.

use crate::card::CardDb;
use crate::state::{GameState, PokemonSlot, get_slot_mut};
use crate::actions::SlotRef;
use crate::types::{CardKind, Stage};
use crate::effects::EffectContext;
use crate::effects::dispatch::apply_effects;

/// Play a Basic Pokemon from hand to a bench slot.
pub fn play_basic(
    state: &mut GameState,
    db: &CardDb,
    hand_index: usize,
    bench_slot: usize,
) {
    let current = state.current_player;
    let player = &state.players[current];

    assert!(
        hand_index < player.hand.len(),
        "Invalid hand_index {}: hand has {} cards",
        hand_index,
        player.hand.len()
    );

    let card_idx = player.hand[hand_index];
    let card = db.get_by_idx(card_idx);

    assert_eq!(card.kind, CardKind::Pokemon, "Card {:?} is not a Pokemon", card.name);
    assert_eq!(
        card.stage,
        Some(Stage::Basic),
        "Card {:?} is not a Basic Pokemon (stage: {:?})",
        card.name,
        card.stage
    );

    assert!(bench_slot < 3, "Invalid bench_slot {}", bench_slot);
    assert!(
        state.players[current].bench[bench_slot].is_none(),
        "Bench slot {} is already occupied",
        bench_slot
    );

    let hp = card.hp;
    // Remove from hand
    state.players[current].hand.remove(hand_index);
    // Place in bench
    state.players[current].bench[bench_slot] = Some(PokemonSlot::new(card_idx, hp));
}

/// Play an Item card: discard it and apply its effects.
///
/// `extra_hand_index` is used by Rare Candy to point to the Stage 2 card.
pub fn play_item(
    state: &mut GameState,
    db: &CardDb,
    hand_index: usize,
    target: Option<SlotRef>,
    extra_hand_index: Option<usize>,
) {
    let current = state.current_player;
    let player = &state.players[current];

    assert!(
        hand_index < player.hand.len(),
        "Invalid hand_index {}: hand has {} cards",
        hand_index,
        player.hand.len()
    );

    let card_idx = player.hand[hand_index];
    let card = db.get_by_idx(card_idx);
    assert_eq!(card.kind, CardKind::Item, "Card {:?} is not an Item", card.name);

    // Capture the extra card index BEFORE popping the item from hand,
    // so indices stay stable.
    let extra_card_idx: Option<u16> = extra_hand_index.and_then(|ei| {
        if ei < state.players[current].hand.len() {
            Some(state.players[current].hand[ei])
        } else {
            None
        }
    });

    // Pop item from hand, discard it.
    state.players[current].hand.remove(hand_index);
    state.players[current].discard.push(card_idx);

    // Adjust extra_hand_index after the pop: if extra came after item, shift down by 1.
    let adjusted_extra_idx: Option<usize> = extra_hand_index.map(|ei| {
        if ei > hand_index { ei - 1 } else { ei }
    });

    // Apply the item card's pre-parsed effects.
    let item_effects = card.trainer_effects.clone();
    let mut ctx = EffectContext {
        acting_player: current,
        source_ref: None,
        target_ref: None,
        extra: Default::default(),
    };
    if let Some(evo_idx) = extra_card_idx {
        ctx.extra.insert("evo_card_idx".to_string(), evo_idx as i32);
    }
    if let Some(adj_idx) = adjusted_extra_idx {
        ctx.extra.insert("evo_hand_index".to_string(), adj_idx as i32);
    }
    if let Some(t) = target {
        let slot_enc: i32 = if t.is_active() { 0 } else { t.slot as i32 + 1 };
        ctx.extra.insert("target_slot".to_string(), t.player as i32 * 10 + slot_enc);
    }

    apply_effects(state, db, &item_effects, &ctx);
}

/// Play a Supporter card: discard, mark flag, apply effects.
pub fn play_supporter(
    state: &mut GameState,
    db: &CardDb,
    hand_index: usize,
    target: Option<SlotRef>,
) {
    let current = state.current_player;
    let player = &state.players[current];

    assert!(
        hand_index < player.hand.len(),
        "Invalid hand_index {}: hand has {} cards",
        hand_index,
        player.hand.len()
    );

    let card_idx = player.hand[hand_index];
    let card = db.get_by_idx(card_idx);
    assert_eq!(
        card.kind,
        CardKind::Supporter,
        "Card {:?} is not a Supporter",
        card.name
    );

    // Pop from hand, discard, mark flag.
    state.players[current].hand.remove(hand_index);
    state.players[current].discard.push(card_idx);
    state.players[current].has_played_supporter = true;

    // Apply the supporter card's pre-parsed effects.
    let supporter_effects = card.trainer_effects.clone();
    let mut ctx = EffectContext {
        acting_player: current,
        source_ref: None,
        target_ref: None,
        extra: Default::default(),
    };
    if let Some(t) = target {
        let slot_enc: i32 = if t.is_active() { 0 } else { t.slot as i32 + 1 };
        ctx.extra.insert("target_slot".to_string(), t.player as i32 * 10 + slot_enc);
    }

    apply_effects(state, db, &supporter_effects, &ctx);
}

/// Attach a Tool card to a Pokemon.
pub fn attach_tool(
    state: &mut GameState,
    db: &CardDb,
    hand_index: usize,
    target: SlotRef,
) {
    let current = state.current_player;
    let player = &state.players[current];

    assert!(
        hand_index < player.hand.len(),
        "Invalid hand_index {}: hand has {} cards",
        hand_index,
        player.hand.len()
    );

    let card_idx = player.hand[hand_index];
    let card = db.get_by_idx(card_idx);
    assert_eq!(card.kind, CardKind::Tool, "Card {:?} is not a Tool", card.name);

    {
        let slot = get_slot_mut(state, target).expect("No Pokemon at target slot");
        assert!(
            slot.tool_idx.is_none(),
            "Target Pokemon already has a tool attached"
        );
        slot.tool_idx = Some(card_idx);
    }

    // Pop from hand (do NOT go to discard — tool stays attached).
    state.players[current].hand.remove(hand_index);

    // Apply the tool card's pre-parsed passive effects.
    let tool_effects = card.trainer_effects.clone();
    let mut ctx = EffectContext {
        acting_player: current,
        source_ref: None,
        target_ref: None,
        extra: Default::default(),
    };
    // Encode target using the effects-system convention:
    // target_slot = player*10 + (bench_idx+1), or 0 for active slot.
    let slot_enc: i32 = if target.is_active() {
        0
    } else {
        target.slot as i32 + 1
    };
    ctx.extra.insert("target_slot".to_string(), target.player as i32 * 10 + slot_enc);

    apply_effects(state, db, &tool_effects, &ctx);
}

// ------------------------------------------------------------------ //
// Tests
// ------------------------------------------------------------------ //

#[cfg(test)]
mod tests {
    use super::*;
    use crate::card::CardDb;
    use crate::state::GameState;
    use crate::types::GamePhase;
    use std::path::PathBuf;

    fn assets_dir() -> PathBuf {
        let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        d.pop();
        d.push("assets/cards");
        d
    }

    fn load_db() -> CardDb {
        CardDb::load_from_dir(&assets_dir())
    }

    fn make_state(db: &CardDb) -> GameState {
        let mut state = GameState::new(42);
        state.phase = GamePhase::Main;
        // Give player 0 a Bulbasaur in hand.
        let bulb = db.get_by_id("a1-001").expect("a1-001 not found");
        state.players[0].hand.push(bulb.idx);
        state
    }

    #[test]
    fn play_basic_places_card_on_bench() {
        let db = load_db();
        let mut state = make_state(&db);
        let hand_len_before = state.players[0].hand.len();
        play_basic(&mut state, &db, 0, 0);
        assert_eq!(state.players[0].hand.len(), hand_len_before - 1, "Card should be removed from hand");
        assert!(state.players[0].bench[0].is_some(), "Bench slot 0 should be occupied");
        let slot = state.players[0].bench[0].as_ref().unwrap();
        let bulb = db.get_by_id("a1-001").unwrap();
        assert_eq!(slot.card_idx, bulb.idx);
        assert_eq!(slot.current_hp, bulb.hp);
        assert_eq!(slot.max_hp, bulb.hp);
        assert_eq!(slot.turns_in_play, 0);
    }

    #[test]
    #[should_panic(expected = "Bench slot 0 is already occupied")]
    fn play_basic_occupied_bench_panics() {
        let db = load_db();
        let mut state = make_state(&db);
        let bulb = db.get_by_id("a1-001").unwrap();
        // Pre-occupy slot 0.
        state.players[0].bench[0] = Some(PokemonSlot::new(bulb.idx, bulb.hp));
        play_basic(&mut state, &db, 0, 0);
    }
}
